### DISCUSSION TITLE
image.toObject(['matrix']) is not work?

### DISCUSSION THREAD

**Discussion by sanheng** [fabricjs/fabric.js#7326] [Q&A]
output not include matrix

**Comment by asturur** [comment]
There is no matrix property by default.
What are you doing exactly?

**Reply by sanheng** [reply]
The backend wants to get the matrix to calculate the position of the picture, circle, rect, Polygon, and text more accurately, because they need to be drawn on the pdf, and the coordinate system of java itext is the lower left corner

**Reply by sanheng** [reply]
For example, for circle’s clippath, the back-end takes the coordinates of image.toObject() to calculate there is always a deviation of 1px.

```
{
	type: 'image',
	clipPath: {
		type: 'circle',
	}

}
```

**Reply by asturur** [reply]
What is this matrix you want to have in the backend? Can you show it on the client?

**Reply by sanheng** [reply]
```
<image width="500" height="500" transform="matrix(0.1344 0 0 0.1344 181.6262 352.7609)" />
```

Like the matrix inside, what should I take is transform?

**Comment by sanheng** [comment]
hello, image.toObject() can't get circle rx ry r property.
image.toObject(['rx', 'ry', 'r']) is not work.

![image](https://user-images.githubusercontent.com/13793256/130892801-43755e99-e51b-4312-9ce4-d2226504c3d8.png)

**Reply by i-prikot** [reply]
Hi. Why aren't you reading the documentation? 

http://fabricjs.com/docs/fabric.Circle.html
http://fabricjs.com/fabric-intro-part-1#objects
https://fabricjs.com/fabric-intro-part-3#serialization

**Reply by sanheng** [reply]
Sorry I didn't read the document carefully.

svg file: 
```html
<text xmlns="http://www.w3.org/2000/svg" id="合并组-1-活动时间-7" transform="matrix(1 -0.1474 0 1 128.6431 194.8145)" fill="#FFFFFF" font-family="'FZFEIFTJF_DA--GBK1-0-GBpc-EUC-H'" font-size="10.08px">活动时间：2021年8月1日-8月31日</text>
```

use fabricjs read:
```
console.log(shape.id, shape.toObject())
console.log('_calcTranslateMatrix', shape._calcTranslateMatrix())
console.log('calcOwnMatrix', shape.calcOwnMatrix())
console.log('getSvgTransform', shape.getSvgTransform())
```
![image](https://user-images.githubusercontent.com/13793256/130932157-f9040d95-9592-419f-b4d9-88d3abffc3ef.png)


How do I use fabricjs to read the same value as the transfrom of the text in the svg file?
```css
transform="matrix(1 -0.1474 0 1 128.6431 194.8145)"
```

**Reply by i-prikot** [reply]
Have a look at this method http://fabricjs.com/docs/fabric.Object.html#calcOwnMatrix

**Reply by sanheng** [reply]
_calcTranslateMatrix(), calcTransformMatrix(), calcOwnMatrix() get the same values.

```js
console.log(shape.id, shape.toObject())
console.log('_calcTranslateMatrix', shape._calcTranslateMatrix())
console.log('calcTransformMatrix', shape.calcTransformMatrix())
console.log('calcOwnMatrix', shape?.calcOwnMatrix())
console.log('getSvgTransform', shape.getSvgTransform())
```
![image](https://user-images.githubusercontent.com/13793256/130935866-4a703a06-a6ba-42d5-a167-0bd169e78eb6.png)

i want to use fabricjs to parse svg to get svg attribute, i want to get the corresponding value in the svg, such as the cx, cy, r of crcle, the x, y, width, height of rect, the transform of text or image.
Because Fabricjs is very good and has parsed these values, but I can’t get them.

svg file:
```html
<text xmlns="http://www.w3.org/2000/svg" id="合并组-1-活动时间-7" transform="matrix(1 -0.1474 0 1 128.6431 194.8145)" fill="#FFFFFF" font-family="'FZFEIFTJF_DA--GBK1-0-GBpc-EUC-H'" font-size="10.08px">活动时间：2021年8月1日-8月31日</text>
<g xmlns="http://www.w3.org/2000/svg" id="合并组-1-商户logo-6">
	<defs>
		<rect id="SVGID_5_" x="242.8" y="529.4" width="113.8" height="36.2"/>
	</defs>
	<clipPath id="SVGID_6_">
		<use xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="#SVGID_5_" overflow="visible"/>
	</clipPath>
	<g clip-path="url(#SVGID_6_)">
		<image overflow="visible" width="474" height="151" xmlns:xlink="base64" transform="matrix(0.24 0 0 0.24 242.775 529.36)">
		</image>
	</g>
</g>
<g xmlns="http://www.w3.org/2000/svg">
	<defs>
		<circle id="SVGID_3_" cx="84.5" cy="311" r="14.9"/>
	</defs>
	<clipPath id="SVGID_4_">
		<use xmlns:xlink="http://www.w3.org/1999/xlink" xlink:href="#SVGID_3_" overflow="visible"/>
	</clipPath>
	<g clip-path="url(#SVGID_4_)">
		<image overflow="visible" width="297" height="297" id="商户1-1-商户图像-1" xmlns:xlink="base64" transform="matrix(0.1068 0 0 0.1068 68.6753 295.2848)">
		</image>
	</g>
</g>
```

**Reply by asturur** [reply]
calcOwnMatrix and calcTransformatrix get the same value, unless the object is inside a group
getSvgTransform return the string for the svg export.

**Reply by sanheng** [reply]
So I can't get the original attributes of element in svg through fabricjs?

such as the cx, cy, r of crcle, the x, y, width, height of rect, the transform of text or image.

**Reply by asturur** [reply]
Everything that is necessary to render the object as in canvas in SVG is added to SVG, with SVG attributes.
Those attributes may have a different names.
the circle CX and CY becomes LEFT, TOP properly translated.

**Reply by asturur** [reply]
The transform="matrix()" from SVG is divided between top, left, scaleX, scaleY, skewX, angle, flipX, flipY
